File: sdk/go/worker_trust_codec.go
package capsdk

import (
	"fmt"

	agentv1 "github.com/cordum-io/cap/v2/cordum/agent/v1"
	"google.golang.org/protobuf/proto"
)

// MarshalWorkerTrustPacket validates and deterministically encodes one phase.
func MarshalWorkerTrustPacket(packet *agentv1.BusPacket) ([]byte, error) {
	if err := ValidateWorkerTrustPacket(packet); err != nil {
		return nil, err
	}
	data, err := proto.MarshalOptions{Deterministic: true}.Marshal(packet)
	if err != nil {
		return nil, fmt.Errorf("capsdk: marshal worker handshake: %w", err)
	}
	if len(data) > WorkerHandshakeMaxPacketSize {
		return nil, fmt.Errorf("%w: exceeds %d bytes", ErrWorkerHandshakePacket, WorkerHandshakeMaxPacketSize)
	}
	return data, nil
}

// UnmarshalWorkerTrustPacket bounds input and rejects unknown fields.
func UnmarshalWorkerTrustPacket(data []byte) (*agentv1.BusPacket, error) {
	if len(data) == 0 || len(data) > WorkerHandshakeMaxPacketSize {
		return nil, fmt.Errorf("%w: size must be 1..%d", ErrWorkerHandshakePacket, WorkerHandshakeMaxPacketSize)
	}
	packet := &agentv1.BusPacket{}
	if err := (proto.UnmarshalOptions{DiscardUnknown: false}).Unmarshal(data, packet); err != nil {
		return nil, fmt.Errorf("capsdk: unmarshal worker handshake: %w", err)
	}
	if err := ValidateWorkerTrustPacket(packet); err != nil {
		return nil, err
	}
	return packet, nil
}

File: sdk/python/cap/worker_trust_codec.py
"""Bounded deterministic codec for worker trust-handshake packets."""

from cap.pb.cordum.agent.v1 import buspacket_pb2
from cap.worker_trust import WORKER_HANDSHAKE_MAX_PACKET_SIZE, WorkerHandshakePacketError
from cap.worker_trust_validate import validate_worker_trust_packet


def marshal_worker_trust_packet(packet: buspacket_pb2.BusPacket) -> bytes:
    """Validate and deterministically encode one trust phase."""

    validate_worker_trust_packet(packet)
    data = packet.SerializeToString(deterministic=True)
    if len(data) > WORKER_HANDSHAKE_MAX_PACKET_SIZE:
        raise WorkerHandshakePacketError("packet size exceeds 65536 bytes")
    return data


def unmarshal_worker_trust_packet(data: bytes) -> buspacket_pb2.BusPacket:
    """Reject empty, oversized, malformed, unknown-field, or invalid input."""

    if not isinstance(data, bytes) or not 0 < len(data) <= WORKER_HANDSHAKE_MAX_PACKET_SIZE:
        raise WorkerHandshakePacketError("packet size must be 1..65536 bytes")
    packet = buspacket_pb2.BusPacket()
    try:
        consumed = packet.ParseFromString(data)
    except Exception as exc:
        raise WorkerHandshakePacketError("cannot decode worker trust packet") from exc
    if consumed is not None and consumed != len(data):
        raise WorkerHandshakePacketError("worker trust packet has trailing bytes")
    validate_worker_trust_packet(packet)
    return packet


__all__ = ["marshal_worker_trust_packet", "unmarshal_worker_trust_packet"]
